### PR TITLE
Prune Validator Attester Records Older than Weak Subjectivity Periods

### PR DESCRIPTION
**Pull request by rauljordan**
Follow-up to #8212, we should be pruning, on startup of the validator client, attestation records older than the weak subjectivity period to maintain the performance of attester slashing protection.

This routine takes around half a second for 1000 validators to prune 54,000 epochs on startup, which is an acceptable trade-off.

### CODE REVIEW

File: validator/db/kv/prune_attester_protection.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		if targetEpoch < wssPeriod {
+			return sourceEpochsBucket.Delete(k)
+		} else if (targetEpoch / wssPeriod) < totalWssPeriods {
+			return sourceEpochsBucket.Delete(k)
+		}
+		return nil
+	})
+}
+
+func pruneSigningRootsBucket(bucket *bolt.Bucket) error {
```
Comment (rauljordan) [inline validator/db/kv/prune_attester_protection.go:67]: Although these functions look similar, the differences are quite subtle and important, which is why we cannot abstract into a single func

File: validator/db/kv/attester_protection.go [hunk 2]
```diff
@@ -42,7 +43,8 @@ func (store *Store) CheckSlashableAttestation(
 		// First we check for double votes.
 		signingRootsBucket := pkBucket.Bucket(attestationSigningRootsBucket)
 		if signingRootsBucket != nil {
-			targetEpochBytes := bytesutil.Uint64ToBytesBigEndian(att.Data.Target.Epoch)
+			targetEpoch := att.Data.Target.Epoch % params.BeaconConfig().WeakSubjectivityPeriod
+			targetEpochBytes := bytesutil.Uint64ToBytesBigEndian(targetEpoch)
```
Comment (prestonvanloon) [inline validator/db/kv/attester_protection.go:47]: Was there a migration for this?

Comment (prestonvanloon) [reply validator/db/kv/attester_protection.go:47]: This doesn't seem right with how pruning works. Can you describe these changes?

**Review by prestonvanloon — CHANGES_REQUESTED** [review]

File: validator/db/kv/attester_protection.go [hunk 3]
```diff
@@ -6,8 +6,9 @@ import (
 	"fmt"
 
 	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
-	"github.com/prysmaticlabs/prysm/shared/bytesutil"
 	bolt "go.etcd.io/bbolt"
+
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
```
Comment (prestonvanloon) [inline validator/db/kv/attester_protection.go:11]: want to revert this file entirely?

**Review by prestonvanloon — DISMISSED** [review]

File: validator/db/kv/prune_attester_protection.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	return sourceEpochsBucket.ForEach(func(k []byte, v []byte) error {
+		targetEpoch := bytesutil.BytesToUint64BigEndian(v)
+
+		// For each source epoch we find, we check
+		// if its associated target epoch is less than the weak
+		// subjectivity period of the highest written target epoch
+		// in the bucket and delete if so.
+		if targetEpoch < wssPeriod {
```
Comment (terencechain) [inline validator/db/kv/prune_attester_protection.go:58]: Consider helper for this

**Review by terencechain — APPROVED** [review]

**Pull request merged**